active/0025-ds-for-retainer.md

```diff
@@ -0,0 +1,185 @@
+
+#### Possible callback implementation
+
+The retainer must provide the following callbacks:
+
+```erlang
+-callback delete_message(context(), topic()) -> ok.
+%% strore_batch([tombstone_message()])
+
+-callback store_retained(context(), message()) -> ok.
+%% strore_batch([message()])
+
+-callback read_message(context(), topic()) -> {ok, list()}.
+%% get_streams for a concrete topic & fold over a concrete topic
+
+-callback page_read(context(), topic(), non_neg_integer(), non_neg_integer()) ->
+    {ok, list()}.
+%% get_streams for a filter & fold over all topics matching filter (!!)
+
+-callback match_messages(context(), topic(), cursor()) -> {ok, list(), cursor()}.
+%% get_streams for a filter & fold over all topics matching filter (!!)
+
+-callback clear_expired(context()) -> ok.
+%% drop generations
+
+-callback clean(context()) -> ok.
+%% drop all generations
+
+-callback size(context()) -> non_neg_integer().
+%% get_streams over all topics & fold (!!)
+```
+
+#### Problems of the straightforward approach
+
+* `page_read` is used from the dashboard and is often used with the `#` topic. Implementing the callback efficiently is impossible — we should fold over _all_ topics, sort somehow, and only then cut out the required page.
+* The same is true for the `size` callback.
+* The same is true for the `match_messages` callback. However, _currently_ we admit that the whole set of retained messages for a subscription is small enough to fit into memory.
+
+### Straightforward approach with optimizations
+
+An obvious optimization is to have slightly different key schemas for the LTS storage of retained messages storage implementation — not
+```
+ts_epoch | lts_index_bits | ts_rest | message_id => message
+```
+
+but just
+
+```
+lts_index_bits | topic => message
+```
+
+to have generation automatically "compacted".
+
+Other optimizations:
+* We do not use generations for retained messages.
+* We implement alternative sharding based on the topic, not on the client id/node id.
+* To simplify replay, we may encapsulate streams for different shards into a single iterator.
+
+Thus each topic is stored uiquely in the storage, and we do not need to fold over all topics to implement `page_read` and `size` callbacks.
+
+With this approach, we may implement `match_messages` callback without folding, in constant space.
+We use iterator state as `context()`.
+
+`page_read` may be implemented in the same manner, however, we need to "scroll" the iterator to the required page.
+
+### Topic indexing approach
+
+The alternative approach is that we use the DS only for _topic indexing_. That is, we store not
+```
+lts_index_bits | topic => message
+```
+but
+```
+lts_index_bits | topic => #message{}
+```
+
+At the same time, we have a separate storage for the messages themselves, and we store them by topic.
+```
+topic => message
+```
+
+The storage
+* is an ordered key-value storage, where the key is a topic, and the value is the message.
```

Comment:
There could be an alternative solution: we could add a special flag to `emqx_ds:open_db/2`, so it automatically deletes the messages from the topic when a new message is published to the topic. But timestamp of the message is still present in the key. If I am not mistaken, it shouldn't introduce too much overhead. 

This has an advantage that the iterators will be incremented when a new message is published, so it should be possible to _subscribe_ to the "retained" topics with minimal modifiction to the `emqx_persistent_session_ds` logic. It may not be particularly useful for the retainer, but this can be used for exchanging session metadata, bypassing `emqx_cm`.

The session can "subscribe" to a special topic (smth. like `$SESSION/clientid/...`) and listen, for example for takeover events. This way we can eliminate one more backchannel needed for the emqx clustering. Also, session state changes can be trivially inspected by subscribing to these special topics. 

Like I said, it's probably useless for the retainer, but if it can reuse the same logic as the session storage manager, it's a win.